policy-controller/k8s/index/src/index.rs

```diff
@@ -0,0 +1,712 @@
+            removed: HashSet<String>,
+        }
+
+        // Aggregate all of the updates by namespace so that we only reindex
+        // once per namespace.
+        let mut updates_by_ns = HashMap::<String, Ns>::default();
+        for srv in srvs.into_iter() {
+            let namespace = srv.namespace().expect("server must be namespaced");
+            let name = srv.name();
+            let server = server::Server::from_resource(srv, &self.cluster_info);
+            updates_by_ns
+                .entry(namespace)
+                .or_default()
+                .added
+                .push((name, server));
+        }
+        for (ns, names) in deleted.into_iter() {
+            updates_by_ns.entry(ns).or_default().removed = names;
+        }
+
+        for (namespace, Ns { added, removed }) in updates_by_ns.into_iter() {
+            if added.is_empty() {
+                self.namespaces.get_with_reindex(namespace, |ns| {
+                    let changed = !removed.is_empty();
+                    for name in removed.into_iter() {
+                        ns.policy.servers.remove(&name);
+                    }
+                    changed
+                });
+            } else {
+                self.namespaces
+                    .get_or_default_with_reindex(namespace, |ns| {
+                        let mut changed = !removed.is_empty();
+                        for name in removed.into_iter() {
+                            ns.policy.servers.remove(&name);
+                        }
+                        for (name, server) in added.into_iter() {
+                            changed = ns.policy.update_server(name, server) || changed;
+                        }
+                        changed
+                    });
+            }
+        }
+    }
+}
+
+impl kubert::index::IndexNamespacedResource<k8s::policy::ServerAuthorization> for Index {
+    fn apply(&mut self, saz: k8s::policy::ServerAuthorization) {
+        let namespace = saz.namespace().unwrap();
+        let name = saz.name();
+        match server_authorization::ServerAuthz::from_resource(saz, &self.cluster_info) {
+            Ok(meta) => self
+                .namespaces
+                .get_or_default_with_reindex(namespace, move |ns| {
+                    ns.policy.update_server_authz(name, meta)
+                }),
+            Err(error) => tracing::error!(%error, "Illegal server authorization update"),
+        }
+    }
+
+    fn delete(&mut self, namespace: String, name: String) {
+        self.namespaces.get_with_reindex(namespace, |ns| {
+            ns.policy.server_authorizations.remove(&name).is_some()
+        })
+    }
+
+    fn reset(
+        &mut self,
+        sazs: Vec<k8s::policy::ServerAuthorization>,
+        deleted: HashMap<String, HashSet<String>>,
+    ) {
+        #[derive(Default)]
+        struct Ns {
+            added: Vec<(String, server_authorization::ServerAuthz)>,
+            removed: HashSet<String>,
+        }
```

Comment:
kind of wonder if this and the similarly-shaped code in the impl for `Server`s could be factored out, but it looks like it's only duplicated twice, so it may not be worth it yet...